fix(evil): Escape and evil-escape do not close Company auto-complete

niontrix opened the pull request:
Why?:
- Pressing Escape or calling evil-escape using "fd" or your configured keybinding does not close Company's auto-complete pop-up especially using LSP. See also: https://github.com/syl20bnr/spacemacs/issues/4242

This change addresses the need by:
- Add hook to call company-abort when entering evil-normal-state

I'm actually not that sure that the code is in the right place. Maybe I should move it to the evil setup in bootstrap.
Can someone point me in the right direction?

Commit:
fix(evil): Escape and evil-escape do not close Company auto-complete

Why?:
- Pressing Escape or calling evil-escape using "fd" or your configured
  keybinding does not close Company's auto-complete pop-up especially using LSP.
  See also: https://github.com/syl20bnr/spacemacs/issues/4242

This change addresses the need by:
- Add hook to call company-abort when entering evil-normal-state

dalanicolai commented on layers/+spacemacs/spacemacs-evil/packages.el:
```diff
@@ -138,6 +138,9 @@
     :config
     (add-hook 'spacemacs-editing-style-hook #'spacemacs//evil-escape-deactivate-in-holy-mode)
+    (add-hook 'evil-normal-state-entry-hook (lambda ()
+                                              (when (and (fboundp 'company--active-p) (company--active-p))
```
I would suggest instead of using `(fbound 'company--active-p)` to wrap the thing in a `(with-eval-after-load 'company ...)` so that the hook is only added when company gets loaded.

niontrix replied:
Thanks for the hint. But I would rather check if `company--active-p` is bound, because it is not a public function and thus could be removed or renamed, which would then cause an error and be a bad UX. What do you think?

pataquets replied:
Both @dalanicolai's and your point are compatible, if I'm not wrong. I see benefit of not adding the hook at all until company gets loaded.

Commit:
fix(evil): Use evil-insert-state-exit-hook

Why?:
- Because someone might from insert-mode to visual-mode and still want the
  Company pop up to be closed.

This change addresses the need by:
- Use evil-insert-state-exit-hook instead of evil-normal-state-entry-hook

syl20bnr commented:
As it has safe guards I would rather put this into evil-escape directly. What do you think ?

smile13241324 commented:
I have moved your fix into evil-escape as this was rather a package issue.
